<launch>
  <!-- 加载雷达 -->
  <include file="$(find lslidar_driver)/launch/lslidar_serial.launch" />
  <!-- 加载机器人模型-->
  <include file="$(find mowen)/launch/display.launch" />
  <include  file="$(find wit_ros_imu)/launch/wit_imu.launch" />

  <!-- 启动rviz -->
  <node pkg="rviz" type="rviz" name="rviz" 
    args="-d $(find nav_demo)/config/gmap.rviz"/>
    <!--轮式里程计-->
    <node pkg="car_bringup" name="pubv" type="pubv.py"  />
    <node pkg="car_bringup" type="base_node" name="odometry_publisher">
      <param name="odom_frame" type="string" value="odom"/>
      <param name="base_footprint_frame" type="string" value="base_footprint"/>
      <param name="linear_scale_x" type="double" value="1.0"/>
      <param name="linear_scale_y" type="double" value="1.0"/>
      <remap from="/sub_vel" to="/vel_raw"/>
      <remap from="/pub_odom" to="/odom_raw"/>
    </node>
   <!--<node pkg="car_bringup" name="odo" type="odo.py"  /> -->
    <!-- 扩展卡尔曼数据融合 -->
    <!-- Extended Kalman data fusion -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" output="screen">
        <remap from="odometry/filtered" to="odom"/>
        <remap from="/imu0" to="/wit/imu"/>
        <remap from="/odom0" to="/odom_raw"/>
        <param name="odom_frame" value="/odom"/>
        <param name="world_frame" value="/odom"/>
        <param name="base_link_frame" value="/base_footprint"/>
        <rosparam command="load" file="$(find car_bringup)/param/robot_localization.yaml"/>
    </node>
<!--
    <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf">
        <param name="output_frame" value="odom"/>
        <param name="base_footprint_frame" value="base_footprint"/>
        <param name="freq" value="30.0"/>
        <param name="sensor_timeout" value="1.0"/>
        <param name="odom_used" value="true"/>
        <param name="imu_used" value="false"/>
        <param name="vo_used" value="false"/>
        <param name="debug" value="false"/>
        <remap from="imu_data" to="/wit/imu" />    
    </node>将节点订阅的imu_data话题改名为imu -->

    <!-- 启动laser_scan_matcher 
    <node pkg="laser_scan_matcher" type="laser_scan_matcher_node" name="laser_scan_matcher_node">
      <param name="fixed_frame" value = "odom"/>
      <param name="max_iterations" value="15"/>
      <param name="base_frame" value="base_footprint"/>
      <param name="use_odom" value="true"/>				
      <param name="publy_pose" value="true"/>
      <param name="publy_tf" value="true"/>
      <param name="use_imu" value="true"/>
      <param name="use_vel" value="true"/>
      <remap from="/vel" to="/vel_raw"/>
      <remap from="/imu/data" to="/wit/imu"/>
    </node>-->



  <!-- 启动gmapping -->
   <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping">
    <param name="map_update_interval" value="1"/> <!-- 地图更新间隔，秒 -->
    <param name="maxUrange" value="6"/> <!-- 最大测距范围，与激光雷达的范围匹配 -->
    <param name="sigma" value="0.05"/> <!-- 激光数据的标准差 -->
    <param name="kernelSize" value="1.0"/> <!-- 高斯核的大小 -->
    <param name="lstep" value="0.05"/> <!-- 激光扫描中的线段步长 -->
    <param name="astep" value="0.05"/> <!-- 激光扫描中的角度步长 -->
    <param name="iterations" value="5.0"/> <!-- 执行的迭代次数 10,5-->
    <param name="lsigma" value="0.075"/> <!-- 激光数据的高斯噪声标准差 -->
    <param name="ogain" value="3.0"/> <!-- 优化增益 -->
    <param name="lskip" value="0.0"/> <!-- 线段跳过率 -->
    <param name="minimumScore" value="4.0"/> <!-- 最小得分，用于评估扫描质量 -->
    <param name="srr" value="0.1"/> <!-- 运动模型的方差（横向运动） -->
    <param name="srt" value="0.2"/> <!-- 运动模型的方差（旋转运动） -->
    <param name="str" value="0.1"/> <!-- 运动模型的方差（前进运动） -->
    <param name="stt" value="0.2"/> <!-- 运动模型的方差（旋转运动） -->
    <param name="linearUpdate" value="0.04"/> <!-- 线性更新步长 -->
    <param name="angularUpdate" value="0.05"/> <!-- 角度更新步长 -->
    <param name="temporalUpdate" value="-1.0"/> <!-- 时间更新步长 -->
    <param name="resampleThreshold" value="0.5"/> <!-- 重采样阈值 -->
    <param name="particles" value="150"/> <!-- 粒子数量，增加粒子数量可以提高精度，但也增加计算负担 100,50-->
    <param name="xmin" value="-7"/> <!-- 地图的最小X坐标 -->
    <param name="ymin" value="-1"/> <!-- 地图的最小Y坐标 -->
    <param name="xmax" value="1"/> <!-- 地图的最大X坐标 -->
    <param name="ymax" value="7"/> <!-- 地图的最大Y坐标 -->
    <param name="delta" value="0.01"/> <!-- 地图分辨率 -->
    <param name="llsamplerange" value="0.01"/> <!-- 激光线段采样范围 -->
    <param name="llsamplestep" value="0.01"/> <!-- 激光线段采样步长 -->
    <param name="lasamplerange" value="0.005"/> <!-- 激光扫描采样范围 -->
    <param name="lasamplestep" value="0.005"/> <!-- 激光扫描采样步长 -->
  </node>
</launch>